File: ios/StableChannels/StableChannels/Models/StableChannel.swift
import Foundation

// MARK: - Bitcoin

struct Bitcoin: Codable, Equatable {
    var sats: UInt64

    static let zero = Bitcoin(sats: 0)

    static func fromSats(_ sats: UInt64) -> Bitcoin {
        Bitcoin(sats: sats)
    }

    static func fromBTC(_ btc: Double) -> Bitcoin {
        let sats = UInt64((btc * Double(Constants.satsInBTC)).rounded())
        return Bitcoin(sats: sats)
    }

    func toBTC() -> Double {
        Double(sats) / Double(Constants.satsInBTC)
    }

    static func fromUSD(_ usd: USD, price: Double) -> Bitcoin {
        let btc = usd.amount / price
        return Bitcoin.fromBTC(btc)
    }

    var formatted: String {
        let btcValue = toBTC()
        return String(format: "%.8f BTC", btcValue)
    }
}

// MARK: - USD

struct USD: Codable, Equatable {
    var amount: Double

    static let zero = USD(amount: 0.0)

    static func fromBitcoin(_ btc: Bitcoin, price: Double) -> USD {
        USD(amount: btc.toBTC() * price)
    }

    func toMsats(price: Double) -> UInt64 {
        let btcValue = amount / price
        let sats = btcValue * Double(Constants.satsInBTC)
        let millisats = sats * 1000.0
        return UInt64(abs(millisats).rounded(.down))
    }

    var formatted: String {
        String(format: "$%.2f", amount)
    }
}

// MARK: - StableChannel

struct StableChannel: Codable {
    var channelId: String  // ldk-node ChannelId is a String in Swift bindings
    var userChannelId: String  // ldk-node UserChannelId is a String in Swift bindings
    var isStableReceiver: Bool
    var counterparty: String  // hex-encoded pubkey (ldk-node PublicKey = String)
    var expectedUSD: USD
    var expectedBTC: Bitcoin
    var stableReceiverBTC: Bitcoin
    var stableProviderBTC: Bitcoin
    var stableReceiverUSD: USD
    var stableProviderUSD: USD
    var riskLevel: Int32
    var timestamp: Int64
    var formattedDatetime: String
    var paymentMade: Bool
    var scDir: String
    var latestPrice: Double
    var prices: String
    var onchainBTC: Bitcoin
    var onchainUSD: USD
    var note: String?
    var nativeChannelBTC: Bitcoin
    var backingSats: UInt64
    var lastStabilityPayment: Int64

    static let `default` = StableChannel(
        channelId: "",
        userChannelId: "",
        isStableReceiver: true,
        counterparty: Constants.defaultLSPPubkey,
        expectedUSD: .zero,
        expectedBTC: .zero,
        stableReceiverBTC: .zero,
        stableProviderBTC: .zero,
        stableReceiverUSD: .zero,
        stableProviderUSD: .zero,
        riskLevel: 0,
        timestamp: Int64(Date().timeIntervalSince1970),
        formattedDatetime: "",
        paymentMade: false,
        scDir: ".data",
        latestPrice: 0.0,
        prices: "",
        onchainBTC: .zero,
        onchainUSD: .zero,
        note: nil,
        nativeChannelBTC: .zero,
        backingSats: 0,
        lastStabilityPayment: 0
    )
}

File: android/app/src/main/java/com/stablechannels/app/models/StableChannel.kt
package com.stablechannels.app.models

import com.stablechannels.app.util.Constants
import kotlinx.serialization.Serializable
import java.util.Locale
import kotlin.math.abs
import kotlin.math.roundToLong

@Serializable
data class Bitcoin(val sats: Long = 0) {
    companion object {
        val ZERO = Bitcoin(0)

        fun fromSats(sats: Long) = Bitcoin(sats)

        fun fromBTC(btc: Double): Bitcoin {
            val sats = (btc * Constants.SATS_IN_BTC).roundToLong()
            return Bitcoin(sats)
        }

        fun fromUSD(usd: USD, price: Double): Bitcoin {
            val btc = usd.amount / price
            return fromBTC(btc)
        }
    }

    fun toBTC(): Double = sats.toDouble() / Constants.SATS_IN_BTC

    val formatted: String
        get() = String.format(Locale.US, "%.8f BTC", toBTC())
}

@Serializable
data class USD(val amount: Double = 0.0) {
    companion object {
        val ZERO = USD(0.0)

        fun fromBitcoin(btc: Bitcoin, price: Double): USD {
            return USD(btc.toBTC() * price)
        }
    }

    fun toMsats(price: Double): Long {
        val btcValue = amount / price
        val sats = btcValue * Constants.SATS_IN_BTC
        val millisats = sats * 1000.0
        return abs(millisats).toLong()
    }

    val formatted: String
        get() = String.format(Locale.US, "$%.2f", amount)
}

@Serializable
data class StableChannel(
    var channelId: String = "",
    var userChannelId: String = "",
    var isStableReceiver: Boolean = true,
    var counterparty: String = Constants.DEFAULT_LSP_PUBKEY,
    var expectedUSD: USD = USD.ZERO,
    var expectedBTC: Bitcoin = Bitcoin.ZERO,
    var stableReceiverBTC: Bitcoin = Bitcoin.ZERO,
    var stableProviderBTC: Bitcoin = Bitcoin.ZERO,
    var stableReceiverUSD: USD = USD.ZERO,
    var stableProviderUSD: USD = USD.ZERO,
    var riskLevel: Int = 0,
    var timestamp: Long = System.currentTimeMillis() / 1000,
    var formattedDatetime: String = "",
    var paymentMade: Boolean = false,
    var scDir: String = ".data",
    var latestPrice: Double = 0.0,
    var prices: String = "",
    var onchainBTC: Bitcoin = Bitcoin.ZERO,
    var onchainUSD: USD = USD.ZERO,
    var note: String? = null,
    var nativeChannelBTC: Bitcoin = Bitcoin.ZERO,
    var backingSats: Long = 0,
    var lastStabilityPayment: Long = 0
) {
    companion object {
        val DEFAULT = StableChannel()
    }
}
